### PR TITLE
[FLIRT] Create signatures from RzAnalysis data

### LINKED ISSUE #817: FLIRT implementation and testing signatures
I've recently implemented FLIRT [elsewhere](https://github.com/williballenthin/lancelot/tree/master/flirt/src) (and [here](https://github.com/williballenthin/viv-utils/blob/master/viv_utils/flirt.py)), and would be happy to share information and test cases from our Apache 2.0 work.

To get started, here is a .sig file (and corresponding .pat file) that contains around 108k FLIRT signatures generated with pcf/sigmake from IDA 7.4 against VC2015/2017/2019 x32/x64 C runtimes and common open source libraries. They are not derived from Hex-Rays signatures and are licensed Apache 2.0 with copyright to FireEye. They should exercise most FLIRT code paths for the latest signature format, including functions larger than u16::MAX.

If rizin wants to rely more on FLIRT, particularly with open source signatures, then do please reach out! We're collecting and preparing a comprehensive signature set for PE analysis and would be happy to have this used elsewhere.
[1.sig.gz](https://github.com/rizinorg/rizin/files/6133304/1.sig.gz)

[1.pat.gz](https://github.com/rizinorg/rizin/files/6133303/1.pat.gz)

### PR DESCRIPTION
**Pull request by wargio**
# squash me

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

1. Implements the creation of FLIRT signatures in `.sig` (compressed) and `.pac` human readable formats
2. Provides customization via `e` 

```
 flirt.node.optimize: FLIRT optimization option when creating a signature file (none: 0, normal: 1, smallest: 2)
   flirt.sig.deflate: enables/disables FLIRT zlib compression when creating a signature file (available only for .sig files)
      flirt.sig.file: FLIRT file list (comma separated) for sig format
   flirt.sig.library: FLIRT library name for sig format
        flirt.sig.os: FLIRT operating system list (comma separated) for sig format
   flirt.sig.version: FLIRT version for sig format
```

3. Implements FLIRT creation via new command `zfc`: example `zfc /my/flirt.pac` or `zfc /my/flirt.sig`
4. Supports `zlib` compression on creation
5. Rewritten rz-sign to support creation, dumping and conversion of FLIRT files and dropped zignatures.

```
Usage: rz-sign [options] [file]
 -h                          this help message
 -a [-a]                     add extra 'a' to analysis command (available only with -o option)
 -e [k=v]                    set an evaluable config variable (available only with -o option)
 -c [output.pac] [input.sig] parses a FLIRT signature and converts it to its other format
 -o [output.sig] [input.bin] performs an analysis on the binary and generates the FLIRT signature.
 -d [flirt.sig]              parses a FLIRT signature and dump its content
 -q                          quiet mode
 -v                          show version information
Examples:
  rz-sign -d signature.sig
  rz-sign -c new_signature.pac old_signature.sig
  rz-sign -o libc.sig libc.so.6
```

**Test plan**

Generation of valid files which can be parsed and applied to rizin and other tools.

How to test the PR:
Open a file (possibly with symbols) run `aa ; zfc /path/to/my/new/flirt.sig` then strip the file and reopen it again, then `aaa ; zfs /path/to/my/new/flirt.sig`

**Closing issues**

Closes #817

### CODE REVIEW

File: librz/core/cconfig.c [hunk 1]
```diff
@@ -2892,6 +2892,15 @@ static bool cb_dbg_verbose(void *user, void *data) {
 	return true;
 }
 
+static bool cb_flirt(void *user, void *data) {
```
Comment (XVilka) [inline librz/core/cconfig.c:2904]: Add `rz_return_val_if_fail()`

File: librz/core/cfile.c [hunk 2]
```diff
@@ -888,25 +889,39 @@ static bool map_multi_dex(RzCore *core, RzIODesc *desc, ut32 id) {
...
 		return true;
 	}
+	RzCoreFile *cf = rz_core_file_cur(core);
+
+	// adds the current size and align next address
+	ut64 base_address = 0x200000;
+	RzBinFile *cur = rz_pvector_empty(&cf->binfiles) ? NULL : rz_pvector_tail(&cf->binfiles);
+	if (cur) {
+		RzIODesc *iod = rz_io_desc_get(core->io, cur->fd);
+		if (iod) {
```
Comment (XVilka) [inline librz/core/cfile.c]: Maybe reverse the condition and switch to the "early exit" pattern?

File: librz/core/cfile.c [hunk 3]
```diff
@@ -888,25 +889,39 @@ static bool map_multi_dex(RzCore *core, RzIODesc *desc, ut32 id) {
 	if (!rz_str_endswith(desc->name, ".dex")) {
 		return true;
 	}
+	RzCoreFile *cf = rz_core_file_cur(core);
+
+	// adds the current size and align next address
+	ut64 base_address = 0x200000;
```
Comment (XVilka) [inline librz/core/cfile.c]: Make that a define maybe?

File: librz/core/cfile.c [hunk 4]
```diff
@@ -1098,14 +1113,18 @@ RZ_API bool rz_core_bin_load(RZ_NONNULL RzCore *r, RZ_NULLABLE const char *filen
...
+RZ_API RZ_BORROW RzCoreFile *rz_core_file_open_many(RZ_NONNULL RzCore *r, RZ_NULLABLE const char *file, int perm, ut64 base_address) {
 	RzList *list_fds = rz_io_open_many(r->io, file, perm, 0644);
 
 	if (rz_list_empty(list_fds)) {
 		rz_list_free(list_fds);
 		return NULL;
 	}
 
+	if (!base_address) {
+		base_address = 0x08000000;
```
Comment (XVilka) [inline librz/core/cfile.c]: Same, move that to a define, I think.

File: librz/core/csign.c [hunk 5]
```diff
@@ -7,6 +7,249 @@
...
+
+	tokens = rz_str_split_duplist(file_list, ",", true);
+	if (!tokens) {
+		RZ_LOG_ERROR("cannot allocate token list\n");
+		return 0;
+	}
+
+	rz_list_foreach (tokens, it, tk) {
+		for (ut32 i = 0; i < RZ_ARRAY_SIZE(file_map); ++i) {
+			if (!strcmp(tk, file_map[i].name)) {
```
Comment (XVilka) [inline librz/core/csign.c]: Reverse the condition to call "continue"

File: librz/core/csign.c [hunk 6]
```diff
@@ -7,6 +7,249 @@
...
+
+	tokens = rz_str_split_duplist(os_list, ",", true);
+	if (!tokens) {
+		RZ_LOG_ERROR("cannot allocate token list\n");
+		return 0;
+	}
+
+	rz_list_foreach (tokens, it, tk) {
+		for (ut32 i = 0; i < RZ_ARRAY_SIZE(os_map); ++i) {
+			if (!strcmp(tk, os_map[i].name)) {
```
Comment (XVilka) [inline librz/core/csign.c]: Same, reverse the condition maybe?

File: librz/signature/flirt.c [hunk 7]
```diff
@@ -1158,6 +935,41 @@ static int parse_v10_header(RzBuffer *buf, idasig_v10_t *header) {
...
+
+	if (rz_buf_read(buffer, &header->version, sizeof(header->version)) != sizeof(header->version)) {
+		goto exit;
+	}
+
+	ret = header->version;
+
+exit:
+	free(header);
+
```
Comment (XVilka) [inline librz/signature/flirt.c]: Unnecessary empty lines.

**Review by XVilka — CHANGES_REQUESTED** [review]

**Comment by XVilka** [comment]
```
FAILED: librz/signature/librz_sign.so.0.4.0-git.p/flirt.c.o 
tcc -Ilibrz/signature/librz_sign.so.0.4.0-git.p -I. -I.. -I../librz/include -Isubprojects/sdb -I../subprojects/sdb -Isubprojects/sdb/src -I../subprojects/sdb/src -I../librz/analysis/arch/gb -I../librz/type/parser -Wall -O2 -g -D_GNU_SOURCE -Werror=sizeof-pointer-memaccess -fvisibility=hidden -DRZ_PLUGIN_INCORE=1 -MD -MF librz/signature/librz_sign.so.0.4.0-git.p/flirt.c.o.d -o librz/signature/librz_sign.so.0.4.0-git.p/flirt.c.o -c ../librz/signature/flirt.c
../librz/signature/flirt.c:670: error: function-like macro 'if' is not defined
```

File: librz/core/cfile.c (hunk 2 above)
Comment (wargio) [reply librz/core/cfile.c]: This condition cannot be reversed because it is an optional one.

**Comment by wargio** [comment]
ASAN here: https://github.com/rizinorg/rizin/pull/2079

**Comment by wargio** [comment]
maybe the sig options should be under `flirt.sig.*`

**Comment by XVilka** [comment]
@wargio I think better the opposite: `sig.flirt.*`

**Comment by wargio** [comment]
> @wargio I think better the opposite: `sig.flirt.*`

i'm referring to the file type options, so `flirt.sig.*` and `flirt.pac.*`

File: librz/core/cconfig.c [hunk 8]
```diff
@@ -3787,6 +3797,24 @@ RZ_API int rz_core_config_init(RzCore *core) {
...
+	SETOPTIONS(n, "msdos", "win", "os2", "netware", "unix", "other", "all", "none", NULL);
+	n = NODECB("flirt.header.os", "all", &cb_flirt);
+	SETDESC(n, "FLIRT operating system list (comma separated) for sig format");
+	SETOPTIONS(n,
+		"aixar", "aout", "ar", "bin", "coff", "dos:com", "dos:com:old", "dos:exe", "dos:exe:old",
+		"dosdrv", "elf", "intelhex", "le", "loader", "lx", "moshex", "ne", "nlm", "omf", "omflib",
+		"pe", "pilot", "srec", "w32run", "zip", "all", "none", NULL);
+	n = NODECB("flirt.header.app", "all", &cb_flirt);
+	SETDESC(n, "FLIRT app list (comma separated) for sig format");
+	SETOPTIONS(n, "console", "graphics", "exe", "dll", "drv", "thread:single", "thread:multi", "16bit", "32bit", "64bit", "all", "none", NULL);
```
Comment (XVilka) [inline librz/core/cconfig.c]: Shouldn't this option be inferred automatically?

File: librz/core/cmd_descs/cmd_zign.yaml [hunk 9]
```diff
@@ -185,6 +185,12 @@ commands:
   - name: zf
     summary: Manage FLIRT signatures
     subcommands:
+      - name: zfc
+        cname: zign_flirt_create
+        summary: Create FLIRT a file
```
Comment (XVilka) [inline librz/core/cmd_descs/cmd_zign.yaml]: Probably better to reword. Completely unclear

File: librz/core/csign.c [hunk 10]
```diff
@@ -7,6 +7,266 @@
...
+	{ "dalvik", RZ_FLIRT_SIG_ARCH_DALVIK },
+	// { "65c816", RZ_FLIRT_SIG_ARCH_65C816 },
+	// { "ad2106x", RZ_FLIRT_SIG_ARCH_AD2106X },
+	{ "arc", RZ_FLIRT_SIG_ARCH_ARC },
+	// { "dsp96k", RZ_FLIRT_SIG_ARCH_DSP96K },
+	// { "m16c", RZ_FLIRT_SIG_ARCH_M16C },
+	{ "pic", RZ_FLIRT_SIG_ARCH_PIC16 },
+	{ "riscv", RZ_FLIRT_SIG_ARCH_RISCV },
+	// { "rl78", RZ_FLIRT_SIG_ARCH_RL78 },
+	// { "s390", RZ_FLIRT_SIG_ARCH_S390 },
```
Comment (XVilka) [inline librz/core/csign.c]: "sysz" in Rizin.

File: librz/core/csign.c [hunk 11]
```diff
@@ -7,6 +7,266 @@
...
+	// { "tms320c3", RZ_FLIRT_SIG_ARCH_TMS320C3 },
+	// { "kr1878", RZ_FLIRT_SIG_ARCH_KR1878 },
+	// { "ad218x", RZ_FLIRT_SIG_ARCH_AD218X },
+	// { "oakdsp", RZ_FLIRT_SIG_ARCH_OAKDSP },
+	// { "tlcs900", RZ_FLIRT_SIG_ARCH_TLCS900 },
+	// { "c39", RZ_FLIRT_SIG_ARCH_C39 },
+	{ "cr16", RZ_FLIRT_SIG_ARCH_CR16 },
+	// { "mn102l00", RZ_FLIRT_SIG_ARCH_MN102L00 },
+	// { "tms320c1x", RZ_FLIRT_SIG_ARCH_TMS320C1X },
+	// { "nec_v850x", RZ_FLIRT_SIG_ARCH_NEC_V850X },
```
Comment (XVilka) [inline librz/core/csign.c]: "v850" is also supported.

File: librz/core/csign.c [hunk 12]
```diff
@@ -7,6 +7,266 @@
...
+	// { "h8500", RZ_FLIRT_SIG_ARCH_H8500 },
+	{ "tricore", RZ_FLIRT_SIG_ARCH_TRICORE },
+	// { "dsp56k", RZ_FLIRT_SIG_ARCH_DSP56K },
+	// { "c166", RZ_FLIRT_SIG_ARCH_C166 },
+	// { "st20", RZ_FLIRT_SIG_ARCH_ST20 },
+	// { "ia64", RZ_FLIRT_SIG_ARCH_IA64 },
+	// { "i960", RZ_FLIRT_SIG_ARCH_I960 },
+	// { "f2mc", RZ_FLIRT_SIG_ARCH_F2MC },
+	// { "tms320c54", RZ_FLIRT_SIG_ARCH_TMS320C54 },
+	// { "tms320c55", RZ_FLIRT_SIG_ARCH_TMS320C55 },
```
Comment (XVilka) [inline librz/core/csign.c]: Supported by Rizin as well

File: librz/core/csign.c [hunk 13]
```diff
@@ -7,6 +7,266 @@
...
+
+const MapStringValue arch_map[23] = {
+	{ "x86", RZ_FLIRT_SIG_ARCH_386 },
+	{ "z80", RZ_FLIRT_SIG_ARCH_Z80 },
+	// { "i860", RZ_FLIRT_SIG_ARCH_I860 },
+	// { "8051", RZ_FLIRT_SIG_ARCH_8051 },
+	// { "tms", RZ_FLIRT_SIG_ARCH_TMS },
+	{ "6502", RZ_FLIRT_SIG_ARCH_6502 },
+	// { "pdp", RZ_FLIRT_SIG_ARCH_PDP },
+	// { "68k", RZ_FLIRT_SIG_ARCH_68K },
```
Comment (XVilka) [inline librz/core/csign.c]: Supported as well

File: librz/core/csign.c [hunk 14]
```diff
@@ -7,6 +7,266 @@
...
+typedef struct map_string_value_t {
+	const char *name;
+	ut32 value;
+} MapStringValue;
+
+const MapStringValue arch_map[23] = {
+	{ "x86", RZ_FLIRT_SIG_ARCH_386 },
+	{ "z80", RZ_FLIRT_SIG_ARCH_Z80 },
+	// { "i860", RZ_FLIRT_SIG_ARCH_I860 },
+	// { "8051", RZ_FLIRT_SIG_ARCH_8051 },
```
Comment (XVilka) [inline librz/core/csign.c]: Also supported

File: librz/signature/create.c [hunk 15]
```diff
@@ -0,0 +1,458 @@
...
+ */
+
+#include <rz_flirt.h>
+#include <rz_util.h>
+
+extern void module_free(RzFlirtModule *module);
+extern ut16 flirt_crc16(const ut8 *data_p, size_t length);
+
+#if 0
+static void sig_dbg_node(const char *name, const RzFlirtNode *node) {
```
Comment (XVilka) [inline librz/signature/create.c]: Maybe remove this?

File: librz/util/list.c [hunk 16]
```diff
@@ -18,6 +18,15 @@ RZ_API RzListIter *rz_list_iter_get_next(RzListIter *list) {
 	return list->n;
 }
 
+RZ_API void *rz_list_iter_get_next_data(RzListIter *list) {
```
Comment (XVilka) [inline librz/util/list.c:29]: Add Doxygen please

File: librz/util/strbuf.c [hunk 17]
```diff
@@ -107,13 +107,18 @@ RZ_API bool rz_strbuf_setbin(RzStrBuf *sb, const ut8 *s, size_t l) {
 	return true;
 }
 
-// TODO: there's room for optimizations here
 RZ_API bool rz_strbuf_slice(RzStrBuf *sb, int from, int len) {
```
Comment (XVilka) [inline librz/util/strbuf.c]: Doxygen.

File: test/db/cmd/cmd_zignature [hunk 18]
```diff
@@ -1794,7 +1794,7 @@ aaa
 zfs bins/flirt/elf-x86/libcurl.a.pac
 EOF
 EXPECT=<<EOF
-Found 207 FLIRT signatures via bins/flirt/elf-x86/libcurl.a.pac
+Found 100 FLIRT signatures via bins/flirt/elf-x86/libcurl.a.pac
```
Comment (XVilka) [inline test/db/cmd/cmd_zignature:1797]: Why such a big difference?

File: test/db/cmd/cmd_zignature [hunk 19]
```diff
@@ -1815,6 +1815,6 @@ CMDS=<<EOF
 zfd bins/flirt/elf-x86/libcurl.a.pac~?
 EOF
 EXPECT=<<EOF
-166
```
Comment (XVilka) [inline test/db/cmd/cmd_zignature:1818]: Same

**Review by XVilka — CHANGES_REQUESTED** [review]
Overall is quite good. Awesome work! 💯

File: librz/core/cconfig.c (hunk 8 above)
Comment (wargio) [reply librz/core/cconfig.c]: tbh, we do not use and setting it to all is ok

File: test/db/cmd/cmd_zignature (hunk 18 above)
Comment (wargio) [reply test/db/cmd/cmd_zignature:1797]: because i was parsing the pac file in the wrong way

File: test/db/cmd/cmd_zignature (hunk 19 above)
Comment (wargio) [reply test/db/cmd/cmd_zignature:1818]: same issue

File: librz/core/cconfig.c (hunk 8 above)
Comment (XVilka) [reply librz/core/cconfig.c]: Yes, I think we should remove it.

File: librz/core/cmd_descs/cmd_zign.yaml [hunk 20]
```diff
@@ -185,15 +185,21 @@ commands:
...
+      - name: zfc
+        cname: zign_flirt_create
+        summary: Create a FLIRT file with .pac or .sig
+        args:
+          - name: filename
+            type: RZ_CMD_ARG_TYPE_FILE
       - name: zfd
         cname: zign_flirt_dump
-        summary: Open FLIRT file and dump
+        summary: Opens a FLIRT file (.pac or .sigà) and dumps its contents
```
Comment (XVilka) [inline librz/core/cmd_descs/cmd_zign.yaml]: Before you used "create" now it "opens", please use either. I think it's better to keep "open"

File: librz/core/cmd_descs/cmd_zign.yaml [hunk 21]
```diff
@@ -185,15 +185,21 @@ commands:
...
         args:
           - name: filename
             type: RZ_CMD_ARG_TYPE_FILE
       - name: zfs
         cname: zign_flirt_scan
-        summary: Open FLIRT file and scan
+        summary: Opens a FLIRT file and tries to apply the signatures to the loaded binary
```
Comment (XVilka) [inline librz/core/cmd_descs/cmd_zign.yaml]: Same

**Review by XVilka — APPROVED** [review]

File: librz/core/cconfig.c (hunk 8 above)
Comment (wargio) [reply librz/core/cconfig.c]: you mean, remove the option? i think this option is needed only if you use IDA and anyway setting it to FFFF (i.e. all on rizin) like i have seen in all the other signatures is fine.

Comment (XVilka) [reply librz/core/cconfig.c]: Yes

Comment (wargio) [reply librz/core/cconfig.c]: removed.

**Comment by wargio** [comment]
the question is now, should we squash it?

**Comment by XVilka** [comment]
Yes, squashing is fine.

**Comment by ret2libc** [comment]
@wargio could you please provide more detailed "test instructions" in the first comment of the PR? Since there are a lot of changes i'm having an hard time understanding what i should check.

**Comment by wargio** [comment]
I have updated the test instructions

**Comment by ret2libc** [comment]
```
Indirect leak of 288 byte(s) in 12 object(s) allocated from:                                   
    #0 0x7f238dc2193f in __interceptor_malloc (/lib64/libasan.so.6+0xae93f)                    
    #1 0x7f238d38b34d in rz_list_append ../librz/util/list.c:293                               
    #2 0x7f237f743945 in rz_sign_flirt_node_new ../librz/signature/create.c:426                
    #3 0x7f237dfa8a43 in rz_core_flirt_create_file ../librz/core/csign.c:420                   
    #4 0x7f237e15dd68 in rz_zign_flirt_create_handler ../librz/core/cmd/cmd_zign.c:1423        
```

**Comment by ret2libc** [comment]
```
Direct leak of 32 byte(s) in 1 object(s) allocated from:                                       
    #0 0x7f238dc21af7 in calloc (/lib64/libasan.so.6+0xaeaf7)                                  
    #1 0x7f238d38b0a9 in rz_list_new ../librz/util/list.c:236                                  
    #2 0x7f238d38b0de in rz_list_newf ../librz/util/list.c:249                                 
    #3 0x7f237f742e73 in rz_sign_flirt_node_new ../librz/signature/create.c:368                
    #4 0x7f237dfa8a43 in rz_core_flirt_create_file ../librz/core/csign.c:420                   
    #5 0x7f237e15dd68 in rz_zign_flirt_create_handler ../librz/core/cmd/cmd_zign.c:1423        

```

File: librz/include/rz_core.h [hunk 22]
```diff
@@ -1102,7 +1103,14 @@ RZ_API RzCoreAutocomplete *rz_core_autocomplete_find(RzCoreAutocomplete *parent,
 RZ_API bool rz_core_autocomplete_remove(RzCoreAutocomplete *parent, const char *cmd);
 RZ_API void rz_core_analysis_propagate_noreturn(RzCore *core, ut64 addr);
 
-RZ_API void rz_core_flirt_dump(RZ_NONNULL const char *flirt_file);
+RZ_API void rz_core_flirt_dump_node(RZ_NONNULL const RzFlirtNode *node);
```
Comment (ret2libc) [inline librz/include/rz_core.h]: is really necessary to export this function?

File: test/db/formats/ar [hunk 23]
```diff
@@ -129,150 +129,150 @@ nth paddr      vaddr      bind   type   size lib name
 nth paddr      vaddr      bind   type size lib name             
 ----------------------------------------------------------------
 1   ---------- 0x00000000 LOCAL  FILE 0        arch.c
-2   0x00000040 0x00000040 LOCAL  SECT 0        .text
```
Comment (ret2libc) [inline test/db/formats/ar]: why these changes? What do they have to do with flirt?

Comment (wargio) [reply test/db/formats/ar]: needed to properly map all the .o files inside an .a

Comment (ret2libc) [reply test/db/formats/ar]: how is that related to flirt?

Comment (wargio) [reply test/db/formats/ar]: ok, let me drop the commit on multibin and open them in a different PR

**Review by XVilka — APPROVED** [review]

File: librz/util/list.c [hunk 24]
```diff
@@ -13,16 +13,41 @@ RZ_API void rz_list_iter_free(RzListIter *list) {
...
+ * */
 RZ_API RzListIter *rz_list_iter_get_next(RzListIter *list) {
 	rz_return_val_if_fail(list, NULL);
 	return list->n;
 }
 
+/**
+ * \brief returns the value stored in the next RzList iterator
+ *
+ * */
```
Comment (ret2libc) [inline librz/util/list.c]:
```suggestion
 **/
```

File: librz/include/rz_util/rz_str.h [hunk 25]
```diff
@@ -237,6 +237,7 @@ RZ_API bool rz_str_glob(const char *str, const char *glob);
 RZ_API int rz_str_binstr2bin(const char *str, ut8 *out, int outlen);
 RZ_API char *rz_str_between(const char *str, const char *prefix, const char *suffix);
 RZ_API bool rz_str_startswith(const char *str, const char *needle);
+#define rz_str_startswith_const(b, const_str) (!strncmp(b, const_str, strlen(const_str)))
```
Comment (ret2libc) [inline librz/include/rz_util/rz_str.h]: I think this can be avoided and we would have the same effect by moving the rz_str_startswith definition in the include file with `static inline`. That way *you*, as developer, don't have to know whether the string may or may not be const and we don't have to duplicate essentially the same code as `rz_str_startswith` here.

**Review by ret2libc — COMMENTED** [review]
Please fix just these two comments, then LGTM! Sorry for the delay

File: librz/include/rz_util/rz_str.h (hunk 25 above)
Comment (wargio) [reply librz/include/rz_util/rz_str.h]: let me remove that.

File: librz/util/list.c (hunk 24 above)
Comment (wargio) [reply librz/util/list.c]: i think that is added by clang-format.

Comment (ret2libc) [reply librz/util/list.c]: weird, unless they have changed something very recently. I think it should be `/** **/`
